### PR TITLE
Add readCompact option

### PR DESCRIPTION
**Pull request by Aegeaner**
### Contribution Checklist

Master Issue: #6734

### Motivation

*Pulsar SQL should support readCompact on topics.*

### Modifications

*Add a Split in PulsarSplitManager to handle compacted data, and a option readCompact for user to open it.*

### Verifying this change

- unit tests

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactory.java [hunk 1]
```diff
@@ -160,4 +161,9 @@ void asyncOpenReadOnlyCursor(String managedLedgerName, Position startPosition, M
      */
     void shutdown() throws InterruptedException, ManagedLedgerException;
 
+    /**
+     * Get the bookkeeper.
+     */
+    BookKeeper getBookKeeper();
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedgerFactory.java:167]: We shouldn't expose the BK client on the interface here

**Comment by gaoran10** [comment]
1. I think we couldn't add the config `pulsar.read-compact`, this config is a global config. It is up to users to query the raw data or the compacted data, so we could add metadata for users to add in the SQL statement.
Such as `select price from pulsar.'public/default'.'stock-topic' where __compacted_topic__ = true` is used to query compacted data. Refer to the `PulsarInternalColumn.java`.

2. The compacted data should be read from two parts, one is the compacted ledger and the other are normal ledgers, like the compacted Topic reading. So we should check the topic compacted position first, and generate splits through the compacted ledger and normal ledgers. One split should have one property such as `read-compacted` indicate this split is used to read compacted ledger or normal topic ledgers. We could use the normal topic name to get its compacted ledger name.

@jiazhai @codelipenghui Please confirm.

**Comment by codelipenghui** [comment]
The compacted data is stored in a separate ledger, I think the cursor can't read data from that ledger. You can see https://github.com/apache/pulsar/blob/53407fc598286690790727635bb5067a7ac108e7/pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicImpl.java#L104 .

**Comment by codelipenghui** [comment]
The compacted data is stored in a separate ledger, I think the cursor can't read data from that ledger. You can see https://github.com/apache/pulsar/blob/53407fc598286690790727635bb5067a7ac108e7/pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicImpl.java#L104 .

**Comment by codelipenghui** [comment]
Looks like there are no updates, I will close it first.

**Pull request closed without merging**